src/psyclone/psyir/nodes/array_mixin.py

```diff
@@ -666,77 +694,127 @@ def same_range(self, index: int, array2, index2: int) -> bool:
-        elif self.is_lower_bound(index) or array2.is_lower_bound(index2):
-            # One and only one of self and array2 use the lbound()
-            # intrinsic to specify the lower bound of the array
-            # dimension. In this case assume that the ranges are
-            # different (although they could potentially be the same).
-            return False
-        elif not sym_maths.equal(range1.start, range2.start):
-            # Neither self nor array2 use the lbound() intrinsic to
-            # specify the lower bound of the array dimension. Try to
-            # determine if they are the same by matching the
-            # text. Use symbolic maths to do the comparison.
-            return False
+        # The logic below assumes array expressions come from valid Fortran,
+        # (e.g. a(2:4) = b(2:5) is not valid Fortran)
+        # and therefore, we assume the length of the equivalent ranges in the
+        # same statement matches:
+        # e.g. a(3, :) and b(4:, 4)
+        # a dim 2 and b dim 1 must have the same length, but not necessarely
+        # the same range (lower and upper bounds)
+        n_range1 = len([x for x in self.children[:index]
+                        if isinstance(x, Range)])
+        n_range2 = len([x for x in array2.children[:index2]
+                        if isinstance(x, Range)])
+        assume_same_length = (
+            (self.ancestor(Assignment) is array2.ancestor(Assignment) or
+             self.is_same_array(array2)) and n_range1 == n_range2
+        )
+
+        # Try to get the datatypes, only if they are ArrayType (so we know
+        # that these have the shape attribute)
+        array1_type = None
+        if (isinstance(self, Reference) and isinstance(self.symbol, DataSymbol)
+                and isinstance(self.symbol.datatype, ArrayType)):
+            array1_type = self.symbol.datatype
+        array2_type = None
+        if (isinstance(array2, Reference) and
+                isinstance(array2.symbol, DataSymbol) and
+                isinstance(array2.symbol.datatype, ArrayType)):
+            array2_type = array2.symbol.datatype
+
+        # Try to get the ranges start values
+        range1_start = None
+        range2_start = None
+        # If we have a implicit lower bound, e.g. a(:) = b(:)
+        # we need to prove that they have the same lower bound value on the
+        # declaration. For example
+        #   integer, dimension(1:3) :: a
+        #   integer, dimension(3:5) :: b
+        # would make it "not equal".
+        if self.is_lower_bound(index):
+            if self.is_same_array(array2) and array2.is_lower_bound(index2):
+                return True
+            if not array1_type:
+                return False
+            if array1_type.shape[index] == ArrayType.Extent.DEFERRED:
+                return False
+            if array1_type.shape[index] == ArrayType.Extent.ATTRIBUTE:
+                range1_start = Literal("1", INTEGER_TYPE)
```

Comment:
I do agree that this description is not very clear. For
```
subroutine test(a)
 integer, dimension(:) :: a
 integer, dimension(:), allocatable :: b
```
The 'a' shape is set to ATTRIBUTE and 'b' shape is set to DEFERRED. If a second routine initialises an array as
```
integer, dimension(0:N-1) :: a
call test(a)
```
inside test, it gets remapped to 1->N (or whatever the test dimension shape is, for dimension(2:N+1) it will follow the next condition, but it is fine because the array is reindexed to 2->N+1).  So, I think it's right that for ATTRIBUTE we hardcode it to "1".

However, for 'b' the lower bound depends on the allocate statement, regardless of if its inside or passed by argument. And this is way we need to return False, as we cannot guarantee it without finding and exploring the allocate statement.

I updated the docstring to hopefully make this clear.

Reply:
Aha! Thanks for updating.